### PR TITLE
Push Library.indirect_accessor references up in the call stack

### PR DESCRIPTION
**Pull request by SkySkimmer**
overlays:
- https://github.com/mattam82/Coq-Equations/pull/586
- https://github.com/coq-community/paramcoq/pull/121
- https://github.com/Lysxia/coq-simple-io/pull/66
- https://github.com/QuickChick/QuickChick/pull/358
- https://github.com/ejgallego/coq-serapi/pull/401
- https://github.com/ejgallego/coq-lsp/pull/659
- https://github.com/coq-community/vscoq/pull/756
- https://github.com/MetaCoq/metacoq/pull/1074

### CODE REVIEW

File: plugins/extraction/g_extraction.mlg [hunk 1]
```diff
@@ -60,6 +60,8 @@ let pr_language = function
   | Scheme -> str "Scheme"
   | JSON -> str "JSON"
 
+let access = Library.indirect_accessor[@@warning "-3"]
```
Comment (SkySkimmer) [inline plugins/extraction/g_extraction.mlg:63]: This is not very satisfying IMO, I'll probably wait for a better solution before doing overlays (probably after https://github.com/coq/coq/pull/18421 is merged)

Comment (ejgallego) [reply plugins/extraction/g_extraction.mlg:63]: I'm OK with this, but indeed it'd be good to head towards a more principled solution.

Comment (SkySkimmer) [reply plugins/extraction/g_extraction.mlg:63]: TBH I don't want to do this for the overlays so this PR is waiting for the more principled solution.

**Comment by ejgallego** [comment]
Happy to help with this (overlays and vernac types) if you need a hand @SkySkimmer

**Comment by SkySkimmer** [comment]
I'm not sure about the overlay for quickchick `tactic_quickchick`, the tactic `quickchick_goal` calls extraction which wants to pierce opacity.
If that's something we want to allow tactics to do then using vernactypes to control opaque access isn't correct.

**Comment by ejgallego** [comment]
> I'm not sure about the overlay for quickchick `tactic_quickchick`, the tactic `quickchick_goal` calls extraction which wants to pierce opacity.

Very interesting, I guess we need to think about this.

Tools like QC could indeed hugely benefit if we did provide something like Isabelle's `Task` API, as indeed, there is more than just accessing the opaque stuff, it needs to compile stuff, etc... So a full OCaml runtime access, which we don't provide helpers to do so.

> If that's something we want to allow tactics to do then using vernactypes to control opaque access isn't correct.

What do you mean by "correct" ?

We could refine the extension type to include this case right? (I'm not saying that's gonna be pleasant)

Actually we already wrap the `typed_vernac` stuff in `VernacExtend`, so tactics could declare this dependency just fine.

It is true we are missing still some stuff as to be able to test this finer-grained dep in Flèche (for example), but in the Flèche model there is absolutely no problem in discovering that a tactic depends on opaque stuff just before executing it.

**Comment by SkySkimmer** [comment]
>What do you mean by "correct" ?

I mean every tactic would need to declare that it maybe accesses opaques, so the vernactypes system is useless.

>It is true we are missing still some stuff as to be able to test this finer-grained dep in Flèche (for example), but in the Flèche model there is absolutely no problem in discovering that a tactic depends on opaque stuff just before executing it.

It's not before executing, you need to actually execute it IMO.

For instance
~~~coq
Ltac foo := idtac.

Ltac foo ::= quickchick_goal.

Goal True.
match goal with |- False => foo end.
(* what TypedVernac? *)
~~~
and that's assuming we somehow forbid it when called from ltac in constr or Hint Extern, because then every command that has constrs needs to be able to pierce opaques.

**Comment by SkySkimmer** [comment]
>Actually for coq-lsp / Flèche there would be no problem on having tactics declare dynamic deps (deps that appear before the execution), but while I'd love to have a system like that in Coq (and I think this is what Rust folks have), that's not gonna happen soon.

That sounds interesting, maybe we should work in that direction rather than keep trying to make static deps work.

**Comment by ejgallego** [comment]
The main issue with dynamic deps is that they do require a way to suspend a tactic, or else things become a PITA.

Let's do the indirect accessor example:
```
qc (lemma) := read_proof lemma >>= \proof -> compile_and_run proof
```
so when `read_proof` is called, 2 things can happen:
- the proof for `lemma` was checked, so we can just return it and continue
- the proof for `lemma` was skipped, it is not ready! We need to suspend the continuation, check the missing proof, and _resume with an updated context_. How to do this is a good question, actually I'd love to get that kind of integration, but seems like a lot of work.

Maybe just using an `option` type would be OK.

**Comment by ejgallego** [comment]
Now that I think of it, why does QuickChick need to pierce opacity of `Qed` stuff? That seems very weird to me for a QuickCheck tool to do.

**Comment by SkySkimmer** [comment]
IDK but they disable extraction"s "accessed opaque" warnings so it seems deliberate.

**Comment by ejgallego** [comment]
`Access Opaque` was set to `true` by default in a19029dee21f77391d99b780582f82bb3ba3b831 , the commit message references issue #2952

**Comment by ejgallego** [comment]
Ummmm, maybe we should address the original problem with `Function` and turn this setting to `false`.

Let's see what it breaks on CI.

**Comment by Lysxia** [comment]
> Now that I think of it, why does QuickChick need to pierce opacity of Qed stuff?

From the point of view of a QuickChick developer, it's just a warning that shows up when calling extraction so we silence it because we don't know another way do deal with it. What does "piercing opacity" mean? Does extraction not necessarily imply piercing opacity?

Would turning the `quickchick` tactic into a command help?

**Comment by ppedrot** [comment]
> What does "piercing opacity" mean?

From the point of view of most of the system, Qed proofs are essentially treated as if they were Admitted. This ensures that most of the time, commands will behave the same in eager mode (the normal one where proofs are processed sequentially) and in delayed mode (the one you get with -quick and in some user interfaces, where Qed proofs are handled asynchronously). Piercing opacity means that you're introducing a write barrier that requires all currently delayed proofs to be computed so that you can access the body of a Qed proof.

Extraction pierces opacity, for what I believe to be bad reasons. Extracting a Qed proof should be an error, in my opinion, but for historical reasons we have a warning for that instead. If you hide computational content behind a Qed abstraction, you shouldn't be able to get it back. (Note that Prop Qed-proofs are fine because they get erased.)

> Would turning the quickchick tactic into a command help?

Definitely.

**Comment by Lysxia** [comment]
Thanks for the clarification! Then I agree that in principle QuickChick would be fine with disabling opaque access altogether. Looking at our warnings, QuickChick is accessing some opaque ssr constants: `ssrnat.eqnP ssrbool.iffP ssrbool.idP`. Maybe those shouldn't be opaque?

**Comment by herbelin** [comment]
I don't know well this story about "access", so sorry in advance if my question is irrelevant, but if there is a need for an explicit status of pierceable sealness (deliberately avoiding the word "opacity" to avoid the confusion with the `Opaque` command), maybe should we introduce one?

**Comment by ejgallego** [comment]
I did some research in #18851 , it seems that indeed we really have no case where extraction needs to access opaque constants except for an accident.

IMHO it would be good to forbid this.

**Comment by ejgallego** [comment]
The problem with math-comp is pretty easy to workaround, it is due to lack of normalization on the `eq_op eqNatMixing 3 4` for example, where

`eqNatMixing := { op : nat -> nat -> bool; axiom := eqPn }`

so indeed, the proof is not discarded; there are many ways to solve this, so we should be good.

**Comment by SkySkimmer** [comment]
>QuickChick is accessing some opaque ssr constants: ssrnat.eqnP ssrbool.iffP ssrbool.idP.

All these proof construct a `reflect` 
https://github.com/coq/coq/blob/fc1b697e47f9593b730f105be88e9260edbcb851/theories/Bool/Bool.v#L942-L944
In some sense `reflect` is not computationally interesting, in fact if the inductive was declared in Prop we could still define large elimination for it. But I guess we don't do that because most tactics would not use that definition and so it would be annoying to use.

**Comment by ejgallego** [comment]
> > QuickChick is accessing some opaque ssr constants: ssrnat.eqnP ssrbool.iffP ssrbool.idP.
> 
> All these proof construct a `reflect`
> 
> 
> In some sense `reflect` is not computationally interesting, in fact if the inductive was declared in Prop we could still define large elimination for it. But I guess we don't do that because most tactics would not use that definition and so it would be annoying to use.

@SkySkimmer see my comment above, these proofs are not even used, they appear in the extracted program by accident, ssreflect already packs the computable operator in the structure record, it is just that extraction doesn't reduce the projection and these terms appear spuriously.

**Comment by ejgallego** [comment]
If some extracted proof would really depend on `eqPn` (for example), it wouldn't work as these proofs are not designed to be run.

**Comment by SkySkimmer** [comment]
> it is just that extraction doesn't reduce the projection

Is extracton really expected to do such reductions?

**Comment by ejgallego** [comment]
> > it is just that extraction doesn't reduce the projection
> 
> Is extraction really expected to do such reductions?

I think it is not, so indeed that has to be fixed somehow (it is easy on the user side, or we could add some support)

But the key point is that the `Qed.` ugly proof is not really used.

**Comment by ejgallego** [comment]
So indeed we could imagine a feature that for "canonical" projections would indeed do the inverse process and be sure the projections are computed; but that would require some design / test-cases.

For the time being, it seems reasonable to me that users do replace the needed constants by versions that can be handled.

Need to look at metacoq failure tho, to see if that's related.

**Comment by SkySkimmer** [comment]
metacoq failure looks trivial, the recheck command wants to access opaques to get all dependencies.

**Comment by ejgallego** [comment]
> metacoq failure looks trivial, the recheck command wants to access opaques to get all dependencies.

Thanks for checking; I guess that's a legitimate use of the accessor, as long as it is not in a tactic IMHO we are good.

**Comment by mattam82** [comment]
I agree extraction should not pierce through Qed'ed things. If I knew it was configurable I would have disallowed it in MetaCoq to help us debug cases where our programs could not run inside Coq itself due to computational dependencies on Qed'ed terms.

**Comment by ppedrot** [comment]
We should probably start by turning the warning into an error by default then.

**Comment by ejgallego** [comment]
@SkySkimmer this seems ready to merge?

**Comment by SkySkimmer** [comment]
No, I haven't done the metacoq overlay.
Also are we ok with the current quichick overlay or do we want to turn the tactic into a command before merging?

**Comment by ejgallego** [comment]
> Also are we ok with the current quichick overlay or do we want to turn the tactic into a command before merging?

I'd be happy to try to take care of that on the AccessOpaque PR, but as you prefer.

A way to live with "legacy" users is to have extraction take the accessor as an optional parameter.

If `accesor == None` then we'd assume that AccessOpaque is False (always).

Personally I'm happy to follow the path we (you) consider more efficient.

**Comment by Lysxia** [comment]
The current quickchick overlay is ok. On master I've dropped the dependencies that caused the warning so it's ok if the warning becomes an error now.

**Comment by SkySkimmer** [comment]
metacoq overlay done

**Review by ejgallego — APPROVED** [review]
Great!

**Comment by ejgallego** [comment]
@coqbot: merge now

**Pull request merged**